### PR TITLE
[FSDP] Zero 3 Optimization Support

### PR DESCRIPTION
**Pull request by klshuster**
# Patch description
This PR adds support for zero3 fsdp optimization. Specify via `--ddp-backend zero3`. Zero3 optimization shards not only the optimizer and gradients but also the model weights. This improves memory pressure, and is especially useful for larger models. Note that this can increase latency as there is more communication cost, and as such for smaller models there may be a slight hit to speed compared to zero2 (and model parallel, of course).

Addresses #3753

**NOTE: This requires pytorch >= 1.12 for use**

# Testing steps
- Enabled the zero3 tests already in CI (thanks Stephen!)
- Confirmed that zero3 works when the number of validation examples is not divisible by the number of GPUs
- Confirmed that zero3 works when `--skip-generation False` during training. 
- Conducted a comprehensive analysis of a variety of "staple" models within ParlAI under different distributed settings; see screenshots below

The main conclusions from each of the models below are:
- **Performance** remains roughly the same regardless of model parallel, zero2, or zero3
- **Speed**: for BART, zero2 remains faster. For the other models, we see varying results; ultimately, everything is faster than model parallel
- **GPU Memory**: zero3 

## BART-Large (400M)
Screen Shot 2022-12-01 at 2 53 15 PM

## T5-Large (770M)
Screen Shot 2022-12-01 at 2 51 32 PM

## Reddit 2.7B (base of BlenderBot) 
Screen Shot 2022-12-01 at 2 51 51 PM

## GPT2-XL (1.5B)
Screen Shot 2022-12-01 at 2 52 16 PM

## R2C2 2.7B
Screen Shot 2022-12-01 at 2 52 31 PM

### CODE REVIEW

File: parlai/core/torch_generator_agent.py [hunk 1]
```diff
@@ -516,6 +516,16 @@ def __init__(self, opt: Opt, shared=None):
         else:
             # this is not a shared instance of this class, so do full init
             self.criterion = self.build_criterion()
+
+            def load_init_model() -> Dict[str, Any]:
```
Comment (stephenroller) [inline parlai/core/torch_generator_agent.py:520]: how is this used?

File: parlai/scripts/distributed_train.py [hunk 2]
```diff
@@ -51,8 +52,9 @@ def setup_args(cls):
 
     def run(self):
         with distributed_utils.slurm_distributed_context(self.opt) as opt:
-            self.train_loop = single_train.TrainLoop(opt)
-            return self.train_loop.train()
+            self.train_loop = fsdp_utils.JoinableTrainLoop(opt)
+            with fsdp_utils.fsdp_join(self.train_loop):
```
Comment (stephenroller) [inline parlai/scripts/distributed_train.py:56]: do you need to do this for distributed_eval too?

File: parlai/utils/fsdp.py [hunk 3]
```diff
@@ -7,22 +7,46 @@
...
+    )
 
+    PYTORCH_FSDP_AVAILABLE = True
     FSDP_AVAILABLE = True
 except ImportError:
-    FSDP_AVAILABLE = False
+    PYTORCH_FSDP_AVAILABLE = False
+    try:
+        from fairscale.nn.wrap.auto_wrap import wrap, enable_wrap
+        from fairscale.nn.data_parallel import FullyShardedDataParallel as FSDP
```
Comment (stephenroller) [inline parlai/utils/fsdp.py:42]: why keep around fairscale support?

**Review by stephenroller — COMMENTED** [review]
nice, pretty smooth implementation. i might recommend just axing support for fairscale but you're the boss

**Review by stephenroller — APPROVED** [review]

File: parlai/core/torch_generator_agent.py (hunk 1 above)
Comment (klshuster) [reply parlai/core/torch_generator_agent.py:520]: ahh it's not, that's an artifact. will delete

File: parlai/scripts/distributed_train.py (hunk 2 above)
Comment (klshuster) [reply parlai/scripts/distributed_train.py:56]: yes, forgot about that. also multiprocessing eval

File: parlai/utils/fsdp.py (hunk 3 above)
Comment (klshuster) [reply parlai/utils/fsdp.py:42]: so getting rid of fairscale would force anyone who wants to use distributed training to be on pytorch >=1.12. Is that a reasonable ask, you think?

Comment (stephenroller) [reply parlai/utils/fsdp.py:42]: That's your call but pytorch 1.13 is out (and I'm using it successfully)

Comment (klshuster) [reply parlai/utils/fsdp.py]: i've removed fairscale FSDP

**Comment by klshuster** [comment]
Note: There is currently a bug in pytorch 1.13 that doesn't allow specifying your own pickle module for loading; this is breaking some tests (`test_apex.py`); the fix is merged in https://github.com/pytorch/pytorch/pull/88570, but we'll need to wait for pytorch to patch to allow us to pass tests with pytorch 1.13

for now, will have requirements specify latest 1.12 version
